oscore-cf/src/main/java/org/eclipse/californium/oscore/OSCoreCtx.java

```diff
@@ -0,0 +1,673 @@
+		}
+		
+		String digest = null;
+		switch (this.kdf) {
+		case HKDF_HMAC_SHA_256:
+			digest = "SHA256";
+			break;
+		case HKDF_HMAC_SHA_512:
+			digest = "SHA512";
+			break;
+		case HKDF_HMAC_AES_128:
+		case HKDF_HMAC_AES_256:
+		default:
+			LOGGER.severe("Requested HKDF algorithm is not supported: " + this.kdf.toString());
+			throw new OSException("HKDF algorithm not supported");
+		}
+
+		// Derive sender_key
+		CBORObject info = CBORObject.NewArray();
+		info.Add(this.sender_id);
+		info.Add(this.context_id);
+		info.Add(this.common_alg.AsCBOR());
+		info.Add(CBORObject.FromObject("Key"));
+		info.Add(this.key_length);
+
+		LOGGER.finest("Info sender key: " + bytesToHex(info.EncodeToBytes()));
+
+		try {
+			this.sender_key = deriveKey(this.common_master_secret, this.common_master_salt, this.key_length, digest,
+					info.EncodeToBytes());
+		} catch (CoseException e) {
+			LOGGER.severe(e.getMessage());
+			throw new OSException (e.getMessage());
+		}
+
+		// Derive recipient_key
+		info = CBORObject.NewArray();
+		info.Add(this.recipient_id);
+		info.Add(this.context_id);
+		info.Add(this.common_alg.AsCBOR());
+		info.Add(CBORObject.FromObject("Key"));
+		info.Add(this.key_length);
+
+		LOGGER.finest("Info recipient key: " + bytesToHex(info.EncodeToBytes()));
+
+		try {
+			this.recipient_key = deriveKey(this.common_master_secret, this.common_master_salt, this.key_length, digest,
+					info.EncodeToBytes());
+		} catch (CoseException e) {
+			LOGGER.severe(e.getMessage());
+			throw new OSException (e.getMessage());
+		}
+
+		// Derive common_iv
+		info = CBORObject.NewArray();
+		info.Add(new byte[0]);
+		info.Add(this.context_id);
+		info.Add(this.common_alg.AsCBOR());
+		info.Add(CBORObject.FromObject("IV"));
+		info.Add(this.iv_length);
+
+		LOGGER.finest("Info common IV: " + bytesToHex(info.EncodeToBytes()));
+
+		try {
+			this.common_iv = deriveKey(this.common_master_secret, this.common_master_salt, this.iv_length, digest,
+					info.EncodeToBytes());
+		} catch (CoseException e) {
+			LOGGER.severe(e.getMessage());
+			throw new OSException (e.getMessage());
+		}
+		
+		System.out.println("Common_IV");
+		for(byte b: this.common_iv){
+			System.out.print(Integer.toHexString(b & 0xff) + " ");
+		}
+		System.out.println();
+		
+	}
+
+	/**
+	 * Overrides hasCode to provide a functional implementation for this class.
+	 */
+	@Override
+	public int hashCode() {
+		byte[] c = new byte[sender_id.length + recipient_id.length];
+		System.arraycopy(sender_id, 0, c, 0, sender_id.length);
+		System.arraycopy(recipient_id, 0, c, sender_id.length, recipient_id.length);
+		return ByteBuffer.wrap(c).hashCode();
+	}
+
+	/**
+	 * Overrides equals to provide a functional implementation for this class.
+	 */
+	@Override
+	public boolean equals(Object o) {
+		if (!(o instanceof OSCoreCtx)) {
+			return false;
+		}
+		OSCoreCtx other = (OSCoreCtx) o;
+		return other.sender_id.equals(sender_id) && other.recipient_id.equals(recipient_id);
```

Comment:
I guess `Arrays.equals` is the better choice.

Reply:
I will change this.